prow/git/v2/client_factory.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// ClientFactory knows how to create clientFactory for repos
+type ClientFactory interface {
+	// ClientFromDir creates a client that operates on a repo that has already
+	// been cloned to the given directory.
+	ClientFromDir(org, repo, dir string) (RepoClient, error)
+	// ClientFor creates a client that operates on a new clone of the repo.
+	ClientFor(org, repo string) (RepoClient, error)
+
+	// Clean removes the caches used to generate clients
+	Clean() error
+}
+
+// RepoClient exposes interactions with a git repo
+type RepoClient interface {
+	Publisher
+	Interactor
+}
+
+type repoClient struct {
+	publisher
+	interactor
+}
+
+// NewClientFactory allows for the creation of repository clients
+func NewClientFactory(host string, useSSH bool, username func() (login string, err error), token func() []byte, gitUser func() (name, email string, err error), censor func(content []byte) []byte) (ClientFactory, error) {
+	cacheDir, err := ioutil.TempDir("", "gitcache")
+	if err != nil {
+		return nil, err
+	}
+	var remotes RemoteResolverFactory
+	if useSSH {
+		remotes = &sshRemoteResolverFactory{
+			host:     host,
+			username: username,
+		}
+	} else {
+		remotes = &simpleAuthResolverFactory{
+			host:     host,
+			username: username,
+			token:    token,
+		}
+	}
+	return &clientFactory{
+		cacheDir:   cacheDir,
+		remotes:    remotes,
+		gitUser:    gitUser,
+		censor:     censor,
+		masterLock: &sync.Mutex{},
+		repoLocks:  map[string]*sync.Mutex{},
+		logger:     logrus.WithField("client", "git"),
+	}, nil
+}
+
+// NewLocalClientFactory allows for the creation of repository clients
+// based on a local filepath remote for testing
+func NewLocalClientFactory(baseDir string, gitUser func() (name, email string, err error), censor func(content []byte) []byte) (ClientFactory, error) {
```

Comment:
I already thought that about the initial naming of "localgit": What about renaming this to `NewTestClientFactory` to clarify the purpose? The  `local` part is more of an implementation detail than telling anything about the purpose